### PR TITLE
CRM-21363 Test using ONLY_FULL_GROUP_BY sqlmode as well

### PR DESCRIPTION
**Pull request by seamuslee001**
Overview
----------------------------------------
Ensure that testing occurs using `ONLY_FULL_GROUP_BY` sqlmode if able to ping @totten

---

 * [CRM-21363: Ensure that tests run using ONLY_FULL_GROUP_BY sql_mode for mysql 5.7](https://issues.civicrm.org/jira/browse/CRM-21363)

### CODE REVIEW

**Comment by mfb** [comment]
the default sql_mode on MySQL >= 5.7.8 is ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION

**Comment by totten** [comment]
This is generally a good idea. I share in @mfb's point (although I don't have pithy prescription either.)

Some other notes to address before merging:
 * I'd really like to see the test-results on both `ubu1204` and `ubu1604` beforehand.
 * It's not really NFC because it changes the behavior on all git-based builds.
 * The title/description should probably be updated for

**Comment by seamuslee001** [comment]
@totten @mfb i have refactored it now if you want to check it

**Comment by mfb** [comment]
Is it possible to run the full test suite in 16.04 on this branch?

**Comment by seamuslee001** [comment]
@mfb if @totten can sort out my pw issue then i could do a manual run of it on 1604, it requires someone with ssh access to the box

**Comment by seamuslee001** [comment]
@totten @mfb @lcdservices @xurizaemon i have added in 'ONLY_FULL_GROUP_BY' as well just for test purposes at the moment but may also help see what chaos it brings

**Comment by monishdeb** [comment]
Jenkins test this please

**Comment by mfb** [comment]
hmm I'm not seeing what the failures are here, it's a 404

**Comment by seamuslee001** [comment]
Jenkins test this please

**Comment by seamuslee001** [comment]
@mfb i have just kicked it off again so give it about 45 min and you should see them

**Comment by mfb** [comment]
ok so there are a lot of failures :/ https://test.civicrm.org/job/CiviCRM-Core-PR/17246/consoleText

**Comment by mfb** [comment]
So it seems like it might be trying to use ONLY_FULL_GROUP_BY on mysql 5.5?  But this doesn't exist until 5.7. 

I think what we want to do is run tests with the default sql mode on each version of mysql (5.5 and 5.7).

**Comment by seamuslee001** [comment]
@mfb Mark i think the sql mode is there but it seems to be way way more stricter in 5.5 than in 5.7 for some reason.

**Comment by mfb** [comment]
Ok yes, ONLY_FULL_GROUP_BY does exist but it means something different.  I think what I'm trying to say is that ONLY_FULL_GROUP_BY isn't in the default sql mode until 5.7. 

So, we want to run tests with the default sql mode for each version.  Which includes ONLY_FULL_GROUP_BY for 5.7 but not 5.5

**Comment by seamuslee001** [comment]
@mfb yep so i think we roll back to only setting if if mysql is 5.7 or above.

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by monishdeb** [comment]
@seamuslee001 I reviewed the patch and agree with the patch. However, here's some points I like to add:
1. moving ```supportsFullGroupBy(...)``` to CRM_Utils_SQL ?
2. creating a fn say ```CRM_Utils_SQL::getSqlModes()``` to fetch array of SQL modes
```php
public static function getSqlModes() {
return explode(',', CRM_Core_DAO::singleValueQuery('SELECT @@sql_mode'));
```
3. Adding a protected variable that can be used to decide whether to run/skip ONLY_FULL_GROUP_BY specific test(s). 
```patch
diff --git a/tests/phpunit/CiviTest/CiviUnitTestCase.php b/tests/phpunit/CiviTest/CiviUnitTestCase.php
index 24922f2..94a72e7 100644
--- a/tests/phpunit/CiviTest/CiviUnitTestCase.php
+++ b/tests/phpunit/CiviTest/CiviUnitTestCase.php
@@ -89,6 +89,11 @@ class CiviUnitTestCase extends PHPUnit_Extensions_Database_TestCase {
   protected $tempDirs;
 
   /**
+   * Does the current setup support ONLY_FULL_GROUP_BY mode
+   */
+  protected $_supportFullGroupBy;
+
+  /**
    * @var boolean populateOnce allows to skip db resets in setUp
    *
    *  WARNING! USE WITH CAUTION - IT'LL RENDER DATA DEPENDENCIES
@@ -291,6 +296,8 @@ class CiviUnitTestCase extends PHPUnit_Extensions_Database_TestCase {
     //  Get and save a connection to the database
     $this->_dbconn = $this->getConnection();
 
+    $this->_supportFullGroupBy = CRM_Util_SQL::supportsFullGroupBy();
+
```

**Comment by seamuslee001** [comment]
1. i feel like its consistant with the other functions in CRM_Core_DAO that it should fit here 2. I think that could be ok 3, guess that makes sense

**Comment by seamuslee001** [comment]
@monishdeb decided to go with all your changes have now updated PR

**Comment by monishdeb** [comment]
@seamuslee001 great .. one last thing, is there any JIRA ticket for this?

**Review by monishdeb — APPROVED** [review]

**Comment by seamuslee001** [comment]
@monishdeb no specific one but i figure it could come under the broad banner of CRM-18439 but we may want a specific JIRA ticket

**Comment by seamuslee001** [comment]
@monishdeb tests have passed

**Comment by monishdeb** [comment]
I am happy with the final patch as tested and reviewed from code pov. Test build passes now, JIRA ticket is linked. Hence merging now.

File: tests/phpunit/CiviTest/CiviUnitTestCase.php
Comment (totten) [inline tests/phpunit/CiviTest/CiviUnitTestCase.php]: @monishdeb @seamuslee001  OK, let me see if I follow correctly... this is "adding a protected variable that can be used to decide whether to run/skip ONLY_FULL_GROUP_BY specific test(s)." (I don't see any here now, but maybe this is looking ahead? I'll assume that premise valid.)

The simplest thing would be for those tests to just call `CRM_Utils_SQL::supportsFullGroupBy();` whenever they want, but clearly this does something a bit more complicated. So there must be a reason. Maybe it's a performance concern -- eg *if there's a lot of these tests, then we don't want to run `supportsFullGroupBy()` too often! We'll just do it once during setup.* But if that's the case, then this seems backwards to me. Compare the number of invocations between these:

 1. In the suite of 6000 unit-tests, a handful of specific tests call `supportsFullGroupBy()` and then drop-out.
 2. In the suite of 6000 unit-tests,  every one of them extends the `setUp()` procedure, which always calls `supportsFullGroupBy()`.

So my only theory/rationalization so far is bogus. Can you explain a little better why we need a new property in the base class?

Comment (seamuslee001) [reply tests/phpunit/CiviTest/CiviUnitTestCase.php]: In my mind this doesn't add much extra overhead to the code base, it will allow any test to alter its behavior based on if it can or cannot the full group by mode.  Note it is used now in L390 of `tests/phpunit/CRM/Core/DAOTest.php`

Comment (totten) [reply tests/phpunit/CiviTest/CiviUnitTestCase.php]: That's fine that it doesn't add a much overhead. But *why was done in the first place*? Why does this piece merit special attention in the `setUp()`?

Comment (seamuslee001) [reply tests/phpunit/CiviTest/CiviUnitTestCase.php]: I'll have to leave that to @monishdeb as i followed his advice

Comment (monishdeb) [reply tests/phpunit/CiviTest/CiviUnitTestCase.php]: @totten it was my suggestion, to include this new property to base class in order to make it available for unit tests to run fullGroupBy specific queries. I asked myself "like ```$_dbconn``` what if we add a new property to tell whether the Mysql support fullGroupBy mode or not" but I agree with your point that it would cause an unnecessary burden for tests that extends the ```setUp()```. So I will revert this change in @seamuslee001 PR #11027 or should I create a separate PR for that ?

Comment (monishdeb) [reply tests/phpunit/CiviTest/CiviUnitTestCase.php]: Submitted a new PR https://github.com/civicrm/civicrm-core/pull/11208

**Pull request merged**
